aptos-move/move-examples/swap/sources/coin_wrapper.move

```diff
@@ -0,0 +1,205 @@
+///
+/// The flow works as follows:
+/// 1. Add the coin_wrapper module to the project.
+/// 2. Add a friend declaration for any core modules that needs to call wrap/unwrap. Wrap/Unwrap are both friend-only
+/// functions so external modules cannot call them and leak the internal fungible assets outside of the project.
+/// 3. Add entry functions in the core modules that take coins. Those functions will be calling wrap to create the
+/// internal fungible assets and store them.
+/// 4. Add entry functions in the core modules that return coins. Those functions will be extract internal fungible
+/// assets from the core data structures, unwrap them into and return the coins to the end users.
+///
+/// The fungible asset wrapper for a coin has the same name, symbol and decimals as the original coin. This allows for
+/// easier accounting and tracking of the deposited/withdrawn coins.
+module swap::coin_wrapper {
+    use aptos_framework::account::{Self, SignerCapability};
+    use aptos_framework::aptos_account;
+    use aptos_framework::coin::{Self, Coin};
+    use aptos_framework::fungible_asset::{Self, BurnRef, FungibleAsset, Metadata, MintRef};
+    use aptos_framework::object::{Self, Object};
+    use aptos_framework::primary_fungible_store;
+    use aptos_std::smart_table::{Self, SmartTable};
+    use aptos_std::string_utils;
+    use aptos_std::type_info;
+    use std::string::{Self, String};
+    use std::option;
+
+    // Modules in the same package that need to wrap/unwrap coins need to be added as friends here.
+    friend swap::router;
+
+    const COIN_WRAPPER_NAME: vector<u8> = b"COIN_WRAPPER";
+
+    /// Stores the refs for a specific fungible asset wrapper for wrapping and unwrapping.
+    struct FungibleAssetData has store {
+        // Used during unwrapping to burn the internal fungible assets.
+        burn_ref: BurnRef,
+        // Reference to the metadata object.
+        metadata: Object<Metadata>,
+        // Used during wrapping to mint the internal fungible assets.
+        mint_ref: MintRef,
+    }
+
+    /// The resource stored in the main resource account to track all the fungible asset wrappers.
+    /// This main resource account will also be the one holding all the deposited coins, each of which in a separate
+    /// CoinStore<CoinType> resource. See coin.move in the Aptos Framework for more details.
+    struct WrapperAccount has key {
+        // The signer cap used to withdraw deposited coins from the main resource account during unwrapping so the
+        // coins can be returned to the end users.
+        signer_cap: SignerCapability,
+        // Map from an original coin type (represented as strings such as "0x1::aptos_coin::AptosCoin") to the
+        // corresponding fungible asset wrapper.
+        coin_to_fungible_asset: SmartTable<String, FungibleAssetData>,
+        // Map from a fungible asset wrapper to the original coin type.
+        fungible_asset_to_coin: SmartTable<Object<Metadata>, String>,
```

Comment:
this may not be necessary as we can already distinguish between coin and fungible asset as a coin is a fully qualified struct and a fungible asset is just an address.... need to read more.

Reply:
This is required for the functions below